Refactor functions in string solver

romainbrenguier opened the pull request:
This is to use standard algorithm and simplify the code

romainbrenguier committed:
Refactor gather_indices to use for_each instead of visitor

LAJW reviewed (approved):
Beautiful!

romainbrenguier committed:
Refactor universal_only_in_index to use expression iterators

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2012,31 +2012,16 @@ static optionalt<exprt> find_counter_example(
-  gather_indices_visitort(): indices() {}
-
-  void operator()(const exprt &expr) override
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &expr)
```
The indentation is misleading. It should be clear that the lambda is entirely inside the `for_each` call.

LAJW replied:
I disagree. Indentation is exactly where it should be. The only misleading bit is opening curly brace on a new line - the usual place for it is on the end of the previous line. Unfortunately CBMC linter doesn't allow that kind of indentation.

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2025,31 +2025,17 @@ static array_index_mapt gather_indices(const exprt &expr)
+    [](const exprt &e)
+    {
+      return !e.operands().empty()
+        && (e.id()==ID_plus || e.id()==ID_minus || e.id()==ID_unary_minus);
```
Is that stating that a plus (or minus) expression is non_linear?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2047,33 +2047,14 @@ static bool is_linear_arithmetic_expr(const exprt &expr)
-
-    // If we found the universal variable not in an index_exprt, fail
-    if(e==expr.univ_var() && index_depth==0)
+  for(auto it=expr.depth_begin(); it!=expr.depth_end();)
```
That doesn't seem to do the same thing as the previous implementation. Does it still return true if the univ var is deeply nested in the body?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2047,7 +2047,7 @@ static bool is_linear_arithmetic_expr(const exprt &expr)
 static bool universal_only_in_index(const string_constraintt &expr)
 {
-  for(auto it=expr.depth_begin(); it!=expr.depth_end();)
+  for(auto it=expr.body().depth_begin(); it!=expr.body().depth_end();)
```
Could you please squash this with the previous commit? I still have doubts that it is correct. Can we look at it together?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2003,31 +2003,15 @@ static optionalt<exprt> find_counter_example(
-  expr.visit(v);
-  return v.indices;
+  array_index_mapt indices;
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &expr) { // NOLINT
```
I still find that indentation hard to read. But if I'm the only one in that case then it's fine.

romainbrenguier replied:
@allredj I disabled clang-format on these lines to have brackets on a new line. Can you check it looks ok?

romainbrenguier committed:
Style: Disabling clang-format in get

This is to allow us to put brackets on a new line, as cpplint likes it.

romainbrenguier committed:
Correct constraints added for char_set

The string constraint body should not contain the universal variable
outside of string access.
However this is allowed in the guard, which we use here.
The meaning of the constraint is the same, but it is of a form accepted
by the string solver.
